src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -493,10 +497,32 @@ private function createAutowiredDefinition(\ReflectionClass $typeHint)
 
     private function createTypeNotFoundMessage($type, $label)
     {
-        if (!$classOrInterface = class_exists($type, false) ? 'class' : (interface_exists($type, false) ? 'interface' : null)) {
+        $autowireById = Definition::AUTOWIRE_BY_ID === $this->currentDefinition->getAutowired();
+        if (!$classOrInterface = class_exists($type, $autowireById) ? 'class' : (interface_exists($type, false) ? 'interface' : null)) {
             return sprintf('Cannot autowire %s for service "%s": Class or interface "%s" does not exist.', $label, $this->currentId, $type);
         }
-        $message = sprintf('No services were found matching the "%s" %s and it cannot be auto-registered', $type, $classOrInterface);
+        if (null === $this->types) {
+            $this->populateAvailableTypes();
+        }
+        if (!$autowireById) {
+            $message = sprintf('No services were found matching the "%s" %s and it cannot be auto-registered', $type, $classOrInterface);
+        } elseif (isset($this->ambiguousServiceTypes[$type])) {
+            $message = sprintf('Service "%s" does not exist. This type could be aliased to any of these existing services: "%s"', $type, implode('", "', $this->ambiguousServiceTypes[$type]));
+        } elseif (isset($this->types[$type])) {
+            $message = sprintf('Service "%s" does not exist. This type could be aliased to the existing "%s" service', $type, $this->types[$type]);
+
+            $aliases = array();
+            foreach (array_keys($this->types, $this->types[$type]) as $type) {
+                if ($this->container->has($type)) {
+                    $aliases[] = $type;
+                }
+            }
+            if ($aliases) {
+                $message .= sprintf('. This service is already aliased to "%s"', implode('", "', $aliases));
+            }
```

Comment:
What case is this catching? I don't see anything in `AutowirePassTest` for this.